R/ModelFit-class.R

```diff
@@ -0,0 +1,99 @@
+
+# constructor
+#' Constructs a new ModelFit object
+#'
+#' @param par list of named parameters
+#' @param distribution character name of distribution
+#' @param metric character name of metric
+#' @param value numeric result of applying metric to observed data and fitted distribution
+#' @param density_function function which takes a vector of coordinates, a list of parameters and
+#' a logical determining whether the output should be scaled; returns the fitted distribution density
+#' @param seg_start integer histogram index start of uniform segment
+#' @param seg_end integer histogram index end of uniform segment
+#' @param ... additional parameters, allowing child classes to be built
+#'
+#' @return a Histogram object
+new_ModelFit <- function(
+    par = NULL,
+    distribution = NULL,
+    metric = NULL,
+    value = NULL,
+    density_function = NULL,
+    seg_start = NULL,
+    seg_end = NULL,
+    ...
+  ){
+
+  # Checking types
+  stopifnot(is.list(par) || is.null(par))
+  stopifnot(is.character(distribution))
+  stopifnot(is.character(metric))
+  stopifnot(is.numeric(value) || is.na(value))
+  stopifnot(is.function(density_function))
+  stopifnot(is.integer(seg_start) || is.null(seg_start))
+  stopifnot(is.integer(seg_end) || is.null(seg_start))
+
+  # Creating object
+  x <- list(
+    par = par,
+    dist = distribution,
+    metric = metric,
+    value = value,
+    dens = density_function,
+    seg_start = seg_start,
+    seg_end = seg_end,
+    ...
+  )
+  class(x) <- "ModelFit"
```

Comment:
`ModelFit` is not so specific.. Maybe we can change it to `HZModelFit`? I think this could be a good idea for all of our classes as to avoid clashes in S3 class names.